Add repo sync command

samcoe opened the pull request:
This PR adds the `repo sync` command. The command allows syncing between two repositories, referred to as `destination repository` and `source repository`, where changes from the `source repository` are synced to the `destination repository`. When a `destination repository` is not specified by an argument the current local directory is chosen as the `destination repository`. The command will try to smartly pick the `source repository` based on the `destination repository` but it can also be specified using the `--source` flag. By default the `source repository` default branch will be synced but can be overridden with the `--branch` flag.

Additionally this PR adds in the `counterfeiter` library as an experiment with generating test doubles for interfaces to see if we like it as an alternative to how we are writing tests now. 

Examples syncing a local repo
<img width="895" alt="Screen Shot 2021-06-10 at 10 37 34 AM" src="https://user-images.githubusercontent.com/7969779/121545282-620bee00-c9d8-11eb-97c2-f0a248077af2.png">

Examples syncing a remote repo 
<img width="890" alt="Screen Shot 2021-06-10 at 10 38 39 AM" src="https://user-images.githubusercontent.com/7969779/121545287-633d1b00-c9d8-11eb-8f36-1cba35406b5e.png">

Example trying to sync a remote repo that is not a fork
<img width="895" alt="Screen Shot 2021-06-10 at 10 39 15 AM" src="https://user-images.githubusercontent.com/7969779/121545296-6506de80-c9d8-11eb-8a52-aca10c82e65f.png">

Example trying to sync a remote repo with diverging commits from parent repo
<img width="895" alt="Screen Shot 2021-06-10 at 10 40 09 AM" src="https://user-images.githubusercontent.com/7969779/121545303-66d0a200-c9d8-11eb-86d9-0341678f1d63.png">

Fixes https://github.com/cli/cli/issues/444

samcoe committed:
Add repo sync command

samcoe commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,286 @@
+	srcStr := fmt.Sprintf("%s:%s", ghrepo.FullName(srcRepo), opts.Branch)
+	destStr := fmt.Sprintf("%s:%s", ghrepo.FullName(destRepo), opts.Branch)
+	if local {
+		destStr = fmt.Sprintf(".:%s", opts.Branch)
```
When working on a local repo is `.` a good way to refer to it?

mislav replied:
I don't think this syntax should be presented or otherwise visible to the user. As per my above comment, I would do away with the confirmation prompt entirely.

samcoe commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,286 @@
+	cmd := &cobra.Command{
+		Use:   "sync [<destination-repository>]",
+		Short: "Sync a repository",
+		Long: heredoc.Doc(`
+			Sync destination repository from source repository.
+
+			Without an argument, the local repository is selected as the destination repository.
+			By default the source repository is the parent of the destination repository.
+			The source repository can be overridden with the --source flag.
+		`),
+		Example: heredoc.Doc(`
+			# Sync local repository from remote parent
+			$ gh repo sync
+
+			# Sync local repository from remote parent on non-default branch
+			$ gh repo sync --branch v1
+
+			# Sync remote fork from remote parent
+			$ gh repo sync owner/cli-fork
+
+			# Sync remote repo from another remote repo
+			$ gh repo sync owner/repo --source owner2/repo2
+		`),
```
I think this command has potential to be confusing, do people think this description and examples are clear and explanatory enough?

mislav replied:
I think that the examples are clear enough!

billygriffin commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,286 @@
+		}
+
+		if !fastForward && !opts.Force {
+			return fmt.Errorf("can't sync .:%s because there are diverging commits, try using `--force`", branch)
```
I'm not sure if this is exactly right and might be too verbose, but I'm wary of just suggesting --force without any info about its destructive behavior.
```suggestion
			return fmt.Errorf("can't sync .:%s because there are diverging commits, you can use `--force` to overwrite the commits on .:%s", branch)
```

samcoe committed:
Start tests

samcoe committed:
Extract git interactions into interface

samcoe committed:
Fix up error message

samcoe committed:
Cleanup SyncOptions

samcoe committed:
Add counterfeiter

samcoe committed:
Add tests

samcoe commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,291 @@
+	}
+
+	if dirtyRepo {
+		err = git.Stash([]string{"push"})
```
Does handling stashing in progress work and changing branches have any downsides? Is it too much "magic"?

mislav replied:
Yes, I would suggest avoiding doing any operation on a dirty repository and just bailing out in that case.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,291 @@
+			$ gh repo sync
+
+			# Sync local repository from remote parent on non-default branch
+			$ gh repo sync --branch v1
```
The `--branch` example is interesting to me, because I thought that all branches will be synced that have equivalents on the upstream. But it looks like only 1 branch will ever be synced at a time and can be overridden with `--branch`. Perhaps the overall documentation could include a note about what does "syncing" achieve.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,291 @@
+	cmd.Flags().StringVarP(&opts.SrcArg, "source", "s", "", "Source repository")
+	cmd.Flags().StringVarP(&opts.Branch, "branch", "b", "", "Branch to sync")
+	cmd.Flags().BoolVarP(&opts.Force, "force", "", false, "Discard destination repository changes")
+	cmd.Flags().BoolVarP(&opts.SkipConfirm, "confirm", "y", false, "Skip the confirmation prompt")
```
I had imagined the sync operation to be a safe operation that you routinely invoke to get up-to-date. It feels odd that there would be a `--confirm` flag— what destructive operation is there to confirm? I would rather just specify `--force` to enable non-fast-forwards, but otherwise have the command always just work non-interactively and handle fast-forwards.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,291 @@
+	cs := opts.IO.ColorScheme()
+	if !opts.SkipConfirm && opts.IO.CanPrompt() {
+		if opts.Force {
+			fmt.Fprintf(opts.IO.ErrOut, "%s Using --force will cause diverging commits on %s to be discarded\n", cs.WarningIcon(), destStr)
```
I don't think we should print a warning when you used `--force` on the command line. You must likely know what `--force` does if you explicitly used it.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,291 @@
+	var httpErr api.HTTPError
+	if err != nil && errors.As(err, &httpErr) && notFastForwardErrorMessage.MatchString(httpErr.Message) {
+		destStr := fmt.Sprintf("%s:%s", ghrepo.FullName(destRepo), opts.Branch)
+		return fmt.Errorf("can't sync because there are diverging commits, you can use `--force` to overwrite the commits on %s",
```
Nitpick: could `syncRemoteRepo` return a special error in this case and have its caller handle presenting it in an informative way to the user? That way the API functionality doesn't concern itself too much with presentation

mislav commented on pkg/cmd/repo/sync/git.go:
```diff
@@ -0,0 +1,98 @@
+	if err != nil {
+		return err
+	}
+	return run.PrepareCmd(cmd).Run()
```
Looks like this git functionality now has 3 levels of stubbing:
1. The `gitClient` interface
2. The overridable `gitExecuter.gitCommand` method
3. The `run.PrepareCmd` mechanism

I would argue that only the stubbing via interface is sufficient to isolate git calls from `sync` command functionality.

samcoe replied:
Agreed. I mostly left this in place due to the uncertainty of moving forward with `counterfeiter`.

mislav commented on pkg/cmd/repo/sync/git.go:
```diff
@@ -0,0 +1,98 @@
+	"github.com/cli/cli/internal/run"
+)
+
+//counterfeiter:generate . gitClient
```
I see only 2 of the `gitClient` methods explicitly stubbed in tests, and yet the tests pass and I'm sure that they invoke more than just 2 methods. Does counterfeiter have default return values for methods that were not explicitly stubbed? I find this to be a dealbreaker because I would prefer that tests have to explicitly set up every stub for every invocation of the method.

samcoe replied:
Yes 100%. That is what is happening here and it totally threw me off. I agree, I dislike this behavior and would rather everything blew up if the function is called and is not stubbed.

mislav commented on pkg/cmd/repo/sync/sync_test.go:
```diff
@@ -0,0 +1,434 @@
+					httpmock.StringResponse(`{"data":{"repository":{"defaultBranchRef":{"name": "trunk"}}}}`))
+			},
+			gitStubs: func(fgc *syncfakes.FakeGitClient) {
+				fgc.HasLocalBranchReturns(true)
```
1. Does counterfeiter fail a test if this stubbed method was never invoked?
2. Does counterfeiter fail a test if there were more invocations than stubs for the same method?

These the features I would ideally want from a stubbing solution.

samcoe replied:
1. No
2. No

Counterfeiter does not provide that functionality by default, but gives the user the ability to check those things through the generated interface it creates, so we could write an assertion for the number of times a function was called.

mislav reviewed (changes requested):
I'm fired up about this command!

I would like to see two major simplifications to the command's implementation:
1. Removal of the interactive prompt. I think that the prompt doesn't add much value to the user, but adds much complexity to the command. As far as I'm concerned as a user, syncing should always "just work" unless a sync would result in a loss of data. In that case, it should fail and instruct me to use `--force`.
2. Splitting up the mode of running when we are syncing the local git repo vs. syncing a remote fork with its parent. I feel these are two very different approaches, but I think they are intertwined in code. Having two completely separate implementations could be beneficial for long-term maintenance of this command.

billygriffin commented:
> Removal of the interactive prompt. I think that the prompt doesn't add much value to the user, but adds much complexity to the command. As far as I'm concerned as a user, syncing should always "just work" unless a sync would result in a loss of data. In that case, it should fail and instruct me to use --force.

I'm not 100% sure I agree with this, though I see what you mean. I'm wondering whether for a lot of people if the prompt is just an extra level of confidence that it's going to do what they think it's going to do (or not). Given how confused we all were in talking through all the possible scenarios, I'm just concerned that if it "just works" it raises the potential for uncertainty and frustration. Open to either way, but wanted to raise that for discussion with the team.

samcoe committed:
Revert "Add counterfeiter"

This reverts commit 096f30a3199f1a7cb5227b6abc66c926b06b510c.

samcoe committed:
Rewrite tests with new mocks

samcoe committed:
Address PR comments

samcoe commented:
@billygriffin @vilmibm @mislav At first I was of the mind that the confirmation prompt helped with reducing confusion of the command, and in our offline discussion I said as much, but after addressing the PR comments and working on the documentation for the command more I have flipped my opinion to match @mislav. The confirmation prompt doesn't add a lot of value.

I think when we originally discussed the functionality that this command would encompass, it was confusing because we were envisioning it doing a lot more, but in its current form I think this command is fairly straight forward. Additionally, I think that if we are trying to reduce confusion about what a command is doing a confirmation prompt is not the right solution, but improving docs or reworking the command to do less is more appropriate. Lastly, the command is not doing anything destructive unless the `--force` flag is used, and I am worried about introducing the behavior of confirmations before non-destructive actions since we don't have it anywhere else from what I can tell.

samcoe committed:
Move dirty repo check before any network calls

vilmibm commented:
that works for me.

samcoe committed:
remove now unused color code

vilmibm reviewed (changes requested):
I did some testing and I'm a bit concerned about the reliability of this command; I quickly got into a state that was confusing and erroneous.

## Setup

I have two repositories:

- evilmibm/cackle
- vilmibm/cackle (fork of evilmibm/cackle)

I cloned vilmibm/cackle, setting evilmibm/cackle as the base repo.

I then committed to both parent and fork, ensuring there would be a conflict.

## expected conflict, got false positive

```bash
gh repo sync # allow to sync from evilmibm/cackle
# => this worked fine; local repo matches evilmibm/cackle:master
gh repo sync --source vilmibm/cackle # attempt to sync from the fork
# => claims to have succeeded, but local repo state has not changed
```

## half-completed force sync

I prepared a branch with conflicting changes and saw the expected "can't sync" error; but then when I used force it seemed to work but I do see an opaque git error:

```bash
gh repo sync -bevilmibm-patch-1
can't sync because there are diverging changes, you can use `--force` to overwrite the changes
gh repo sync -bevilmibm-patch-1 --force
exit status 128
git log
# => see that I am on evilmibm-patch-1 branch with expected changes; unsure what the error is but it makes me afraid
gh repo sync
✓ Synced .:master from evilmibm/cackle:master
# => claims to succeed...but still on evilmibm-patch-1 branch. nothing has changed.
```

samcoe committed:
Fix remote resolving for source repo

vilmibm reviewed (changes requested):
Found another issue; branch resolution can fail even with `source` specified when a branch exists in multiple remotes:

![image](https://user-images.githubusercontent.com/98482/123843295-604a9180-d8d7-11eb-830a-eb35bff747e0.png)

samcoe committed:
Catch a couple more edge cases

samcoe commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,322 @@
+			return fmt.Errorf("can't sync because there are diverging changes, you can use `--force` to overwrite the changes")
+		}
+		if errors.Is(err, mismatchRemotesError) {
+			return fmt.Errorf("can't sync because %s is not tracking %s", opts.Branch, ghrepo.FullName(srcRepo))
```
I don't like this error message very much so if there are any suggestions, I am all ears. 

This error can occur when the branch to be synced already exists locally but is pointing to a git remote that does not match the git remote specified by the user in the source argument or the remote we automatically determine using `BaseRepo()`.

vilmibm replied:
It makes sense to me :shrug: I encountered it organically and didn't end up confused.

samcoe commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,322 @@
+				return err
+			}
+		} else {
+			err = git.Checkout([]string{"--track", fmt.Sprintf("%s/%s", remote, branch)})
```
This should ensure that if there are multiple remotes with the same branch name we will checkout the correct one.

mislav replied:
Shouldn't the invocation be: `git checkout <branch> --track <remote>/<branch>`? It looks like the local branch argument is missing

samcoe replied:
According to the docs (https://git-scm.com/docs/git-checkout#Documentation/git-checkout.txt---track) the name to use for the local branch will be automatically determined.

samcoe committed:
Fix lint

vilmibm reviewed (approved):
ran through some QA and everything worked well :sparkles: awesome work

mislav commented on pkg/cmd/repo/sync/git.go:
```diff
@@ -0,0 +1,114 @@
+	return parts[0], nil
+}
+
+func (g *gitExecuter) Checkout(args []string) error {
```
I would guess the primary point of the git abstraction here is to be able to stub it in tests, and for that purpose `gitClient` serves it well.

However, I'd like to imagine the concrete implementation of gitExecuter as something we can pass around and reuse in different commands. In that case, accepting raw cli flags as arguments to methods like `Checkout()` does not feel like sufficient abstraction to me. Would it be a lot of work to rework the git interface to accept arguments that feel more like Go, e.g. string primitives, named arguments, and possibly struct params? That way the caller wouldn't have to know or pass specific git command-line flags.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+		Long: heredoc.Docf(`
+			Sync destination repository from source repository. Syncing will take a branch
+			on the source repository and merge it into the branch of the same name on the
+			destination repository. A fast forward merge will be used execept when the
```
I think using the word "merge" in these sentences might give someone the idea that merge commits will be created, but upon careful reading it sounds like we only ever do fast-forwards (or hard resets in the case of `--force`). Since actual merge commits will never be created, could we avoid using that word and instead use terms like "syncing" or "updating" the branch to match source repo?

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+			this can be overridden with the %[1]s--source%[1]s flag.
+
+			The source repository default branch is selected automatically, but can be
+			overridden with the %[1]s--branch%[1]s flag.
```
This still reads slightly ambiguous. I don't think this paragraph is necessary provided that 1) the single-branch behavior is clarified earlier in the command's docs and 2) the `--branch` flag has sufficient documentation about what it does if it's provided.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+		Short: "Sync a repository",
+		Long: heredoc.Docf(`
+			Sync destination repository from source repository. Syncing will take a branch
+			on the source repository and merge it into the branch of the same name on the
```
“Syncing will take a branch” — could this clarify which branch? Maybe it should point out that, by default, only the main branch is ever synced. Later on, the flags section will point out that `--branch` can be used to select a different branch to be synced.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+			# Sync local repository from remote parent on non-default branch
+			$ gh repo sync --branch v1
+
+			# Sync remote fork from remote parent
```
Small suggestion
```suggestion
			# Sync remote fork from its parent
```

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+	}
+
+	cmd.Flags().StringVarP(&opts.SrcArg, "source", "s", "", "Source repository")
+	cmd.Flags().StringVarP(&opts.Branch, "branch", "b", "", "Branch to sync")
```
Maybe this flag's docs should include a note: `(default: main branch)`

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+
+	cmd.Flags().StringVarP(&opts.SrcArg, "source", "s", "", "Source repository")
+	cmd.Flags().StringVarP(&opts.Branch, "branch", "b", "", "Branch to sync")
+	cmd.Flags().BoolVarP(&opts.Force, "force", "", false, "Discard destination repository changes")
```
This might be slightly ambiguous. How about something like "Hard-reset the branch in the destination repo to math the source repo"?

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+	opts.IO.StopProgressIndicator()
+	if err != nil {
+		if errors.Is(err, divergingError) {
+			return fmt.Errorf("can't sync because there are diverging changes, you can use `--force` to overwrite the changes")
```
Nit: split the sentences
```suggestion
			return fmt.Errorf("can't sync because there are diverging changes; use `--force` to overwrite the destination branch")
```

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+	if opts.IO.IsStdoutTTY() {
+		cs := opts.IO.ColorScheme()
+		srcStr := fmt.Sprintf("%s:%s", ghrepo.FullName(srcRepo), opts.Branch)
+		destStr := fmt.Sprintf(".:%s", opts.Branch)
```
I still think that `.:<something>` is confusing syntax and shouldn't be shown to the user. To my knowledge, this syntax is not encountered anywhere in git nor in GitHub.

Perhaps a more human `Synced <destination> from <source> should be shown` without too many special characters like `:` or `.`? E.g.

```
Synced the "main" branch from upstream to origin

# or:
Synced the "main" branch from owner/repo to local repository
```

samcoe replied:
The fetch and merge feature on GitHub uses `repo:branch` syntax which is why I used it, but I agree using a more human readable message is better.

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+	git := opts.Git
+	branch := opts.Branch
+
+	err := git.Fetch([]string{remote, fmt.Sprintf("+refs/heads/%s", branch)})
```
The `refs/heads/<branch>` ref is fetched into FETCH_HEAD, right? If so, I do not think that `+` is necessary or has any effect. It only has effect in the case of `+refs/heads/<branch>:<destination>`

samcoe replied:
Good catch!

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+	if hasLocalBranch {
+		if opts.Force {
+			err = git.Reset([]string{"--hard", fmt.Sprintf("refs/remotes/%s/%s", remote, branch)})
+			if err != nil {
```
Nitpick: assigning to err and checking for nil can be done in a single line within an `if` condition; I think that might be slightly more idiomatic!

mislav commented on pkg/cmd/repo/sync/sync.go:
```diff
@@ -0,0 +1,325 @@
+		}
+	}
+	if startBranch != branch {
+		err = git.Checkout([]string{startBranch})
```
Isn't this covered by one of the above `git.Checkout`?

samcoe replied:
Not sure I understand the question. The first `git.Checkout` is to checkout the branch to sync and then this `git.Checkout` is to switch back to the initial branch the user was on. I didn't want to assume that the user wanted to switch to the branch that was being sync'ed.

mislav replied:
Ah I've missed that detail. We do not have to switch branches just to update a non-current branch to another ref: there is `git update-ref` for that! The benefit of not switching branches unnecessarily is that the filesystem needs fewer writes (since the working copy is not being updated) and git reflog is cleaner, which benefits tooling. I'll open a PR to address this.

mislav reviewed (changes requested):
Implementation looks good! I suggested a bunch of UI/docs changes that I would ideally like to see before this lands.

One note about the overall functionality: if this commands is designed to sync a single branch from one repository to another, wouldn't `repo sync` be a misnomer? The command name sounds like it sync repositories with each other, but all it does is it sync a single branch with its upstream equivalent. I wonder if the name might mislead people into misinterpreting what this command actually does. But I don't feel very strongly about this concern; just thinking out loud.

samcoe committed:
Clean up UX

samcoe committed:
Use more idiomatic pattern

samcoe committed:
Remove unnecessary +

samcoe committed:
Rework git client interface

samcoe committed:
Fix tests

samcoe commented:
@mislav Thanks for the thorough review! I addressed/commented on your concerns. Regarding the name of the command, do you have better idea? I think `repo branch-sync` might be more accurate but also more verbose.

mislav commented:
> Regarding the name of the command, do you have better idea? I think `repo branch-sync` might be more accurate but also more verbose.

I'm fine with `repo sync` if we in the future expand this command to do more across whole repositories, e.g. sync all local branches to their equivalents? (I'm obviously biased because I that's how I made `hub sync` to work.) Together with another potential feature to clean up branches belonging to merged PRs, I think this command has the potential to be a swiss army knife for repo syncing. But if all it ever does for the foreseeable future is sync a single branch, then `sync-branch` might be a more descriptive command name.

mislav commented on pkg/cmd/repo/sync/git.go:
```diff
@@ -0,0 +1,115 @@
+	return git.HasLocalBranch(branch)
+}
+
+func (g *gitExecuter) IsAncestor(ancestor, progeny string) (bool, error) {
```
Love the argument names 😍

mislav reviewed (approved):
Looks great!

I think we might have discussed this in a meeting in the past, but do you also envision this command being able to sync 3 repos, i.e. local - remote - parent in a single go in the future? I wonder whether people will want to sync their merge branch in their remote fork but also in the local checkout at the same time, and right now it sounds like they will have to make 2 separate `repo sync` invocations.

samcoe commented:
@mislav I don't see this command becoming a swiss army knife type of command and I would like to keep it specific to syncing branches, this could include pruning merged PR branches, and syncing all local branches in the future. I think the main challenge is that the behavior when invoked for a local repo vs a remote repo can get out of sync very fast since we have lots of control over a local repo as opposed to a remote one. In the future when we add features we need to be cognizant of that and think through what the behavior might look like for remote repos or make the features local repo specific. 

We purposefully skipped the feature of syncing 3 repos to see if our users actually wanted it, so you are correct that as of now it will have to be two separate invocation of `repo sync` but I do think that we can introduce that specific case in the future using a flag.

meyerweb commented:
Just tried this out (I’m a little behind) and this is SO GOOD.  Thank you!
